<a href="https://colab.research.google.com/github/HammerMo/ds-oct-23/blob/data-science/Proyecto_I_Sales_Pr_Vicente_Herrera.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyecto 1

Diccionario de datos

`Item_Identifier `- Identificador único del producto

`Item_Weight` - Peso del producto

`Item_Fat_Content` - Si el producto es bajo en grasa o regular

`Item_Visibility` - Porcentaje de área total de exposición de todos los productos para una tienda asignada al producto concreto

`Item_Type` - Categoría a la que el producto pertenece

`Item_MRP ` - Precio máximo de venta al público (precio de catálogo) del producto

`Outlet_Identifier `- Identificador único de la tienda

`Outlet_Establishment_Year` - Año en que se estableció la tienda

`Outlet_Size` - El tamaño de la tienda en relación al área total

`Outlet_Location_Type `- El tipo de área donde se encuentra la tienda

`Outlet_Type` - Si el punto de venta es una tienda de comestibles o algún tipo de supermercado

`Item_Outlet_Sales` - Ventas del producto en una tienda particular. Esta la variable objetivo a predecir.

# Parte 1

In [2]:
# Monto Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
# Importo librerías y archivo
import pandas as pd
import numpy as np
sales=('/content/drive/MyDrive/DS BOOTCAMP/sales_predictions_2023.csv') # carga del archivo
pd.set_option('display.max_columns', None) # muestro todas las columnas

# Parte 2

Exploración de Datos

1) ¿Cuántas filas y columnas hay?

2) ¿Cuáles son los tipos de datos de cada variable?

3) ¿Hay duplicados? Si es el caso, elimínelos.

4) Identifique los valores faltantes.

5) Decida qué hacer con los valores faltantes y cómo hacerlo. (Esto requiere de su criterio, así que explique cada elección).

6) Confirme que no hay valores faltantes después de tratarlos.

7) Encuentre y arregle cualquier categoría inconsistente en los datos (ejemplo: arreglar los datos gato, Gato y Gatos de manera que sean consistentes)

8) Para cualquier columna numérica, obtenga las estadísticas resumidas de cada una (mínimo, máximo y media)

In [4]:
# cargo dataframe
df = pd.read_csv('/content/drive/MyDrive/DS BOOTCAMP/sales_predictions_2023.csv')
# muestro 10 valores y disposicion database para entender que estoy trabajando
df.head(10)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15        9.300          Low Fat         0.016047   
1           DRC01        5.920          Regular         0.019278   
2           FDN15       17.500          Low Fat         0.016760   
3           FDX07       19.200          Regular         0.000000   
4           NCD19        8.930          Low Fat         0.000000   
5           FDP36       10.395          Regular         0.000000   
6           FDO10       13.650          Regular         0.012741   
7           FDP10          NaN          Low Fat         0.127470   
8           FDH17       16.200          Regular         0.016687   
9           FDU28       19.200          Regular         0.094450   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   
5           Baking Goods   51.4008            OUT018   
6            Snack Foods   57.6588            OUT013   
7            Snack Foods  107.7622            OUT027   
8           Frozen Foods   96.9726            OUT045   
9           Frozen Foods  187.8214            OUT017   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   
5                       2009      Medium               Tier 3   
6                       1987        High               Tier 3   
7                       1985      Medium               Tier 3   
8                       2002         NaN               Tier 2   
9                       2007         NaN               Tier 2   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052  
5  Supermarket Type2           556.6088  
6  Supermarket Type1           343.5528  
7  Supermarket Type3          4022.7636  
8  Supermarket Type1          1076.5986  
9  Supermarket Type1          4710.5350

In [5]:
# filas y columnas del df
df.shape

(8523, 12)

EDA - check general data (df.info() - df.shape - df.describe().round(2) -  df.duplicated() - df.duplicated().sum() - id valores perdidos[ df.isna().sum())] - Categorias del dataframe-  df.nunique() - tipos de datos df.type()

In [6]:
# resumen de la data - tipos de datos de cada variable
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [7]:
# exploramos duplicados
df.duplicated().sum()

0

No hay duplicados.

In [8]:
# Valores faltantes
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Existen datos faltantes en:


- `Item_Weight`         1463 , tipo float (pesos)
- `Outlet_Size`         2410 , tipo object (tamaño tienda: small medium large)

Tratamiento de datos para:
- `Item_Weight`       
- `Outlet_Size`        

In [9]:
# Tratamiento Item_Weight
df.isna()['Item_Weight'].sum()

1463

In [10]:
# Generamos un df con la data faltante para ver si es que otra columna nos entrega información que nos falta - no obtenemos nada asique seguimos analizando los nan
df_nan = df['Item_Weight'].isna()
df[df_nan]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10          NaN          Low Fat         0.127470   
18             DRI11          NaN          Low Fat         0.034238   
21             FDW12          NaN          Regular         0.035400   
23             FDC37          NaN          Low Fat         0.057557   
29             FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
8485           DRK37          NaN          Low Fat         0.043792   
8487           DRG13          NaN          Low Fat         0.037006   
8488           NCN14          NaN          Low Fat         0.091473   
8490           FDU44          NaN          Regular         0.102296   
8504           NCN18          NaN          Low Fat         0.124111   

                  Item_Type  Item_MRP Outlet_Identifier  \
7               Snack Foods  107.7622            OUT027   
18              Hard Drinks  113.2834            OUT027   
21             Baking Goods  144.5444            OUT027   
23             Baking Goods  107.6938            OUT019   
29                   Canned   43.6454            OUT019   
...                     ...       ...               ...   
8485            Soft Drinks  189.0530            OUT027   
8487            Soft Drinks  164.7526            OUT027   
8488                 Others  184.6608            OUT027   
8490  Fruits and Vegetables  162.3552            OUT019   
8504              Household  111.7544            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
7                          1985      Medium               Tier 3   
18                         1985      Medium               Tier 3   
21                         1985      Medium               Tier 3   
23                         1985       Small               Tier 1   
29                         1985       Small               Tier 1   
...                         ...         ...                  ...   
8485                       1985      Medium               Tier 3   
8487                       1985      Medium               Tier 3   
8488                       1985      Medium               Tier 3   
8490                       1985       Small               Tier 1   
8504                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
7     Supermarket Type3          4022.7636  
18    Supermarket Type3          2303.6680  
21    Supermarket Type3          4064.0432  
23        Grocery Store           214.3876  
29        Grocery Store           125.8362  
...                 ...                ...  
8485  Supermarket Type3          6261.8490  
8487  Supermarket Type3          4111.3150  
8488  Supermarket Type3          2756.4120  
8490      Grocery Store           487.3656  
8504  Supermarket Type3          4138.6128  

[1463 rows x 12 columns]

In [11]:
# Usamos describe para analizar la estadistica descriptiva de la columna
df['Item_Weight'].describe()

count    7060.000000
mean       12.857645
std         4.643456
min         4.555000
25%         8.773750
50%        12.600000
75%        16.850000
max        21.350000
Name: Item_Weight, dtype: float64

Podemos observar que los elementos de `'Item_Weight'` fluctúan entre 4.55 y 21.35, con una moda de 12.857645.

Evaluando la columna para decidir el tratamiento de datos, nos damos cuenta que: los datos faltantes corresponden a un 17,165%  (1463   ÷   8523) lo que implica que no es un % significativo, por lo que procederemos a reemplazar los valores faltantes con la moda.

In [12]:
# reemplazamos con la moda.
df['Item_Weight'].fillna(df['Item_Weight'].mean(), inplace=True)

In [13]:
# revisamos si es que siguen existiendo valores nulos en 'Item_Weight'
df.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Tratamos correctamente los valores de 'Item_Weight'

In [14]:
# Tratamiento de 'Outlet_Size'
df.isna()['Outlet_Size' ].sum()

2410

In [15]:
# Generamos un df con la data faltante para ver si es que otra columna nos entrega información que nos falta
df_nan = df['Outlet_Size'].isna()
df[df_nan]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07       19.200          Regular         0.000000   
8              FDH17       16.200          Regular         0.016687   
9              FDU28       19.200          Regular         0.094450   
25             NCD06       13.000          Low Fat         0.099887   
28             FDE51        5.925          Regular         0.161467   
...              ...          ...              ...              ...   
8502           NCH43        8.420          Low Fat         0.070712   
8508           FDW31       11.350          Regular         0.043246   
8509           FDG45        8.100          Low Fat         0.214306   
8514           FDA01       15.000          Regular         0.054489   
8519           FDS36        8.380          Regular         0.046982   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
8              Frozen Foods   96.9726            OUT045   
9              Frozen Foods  187.8214            OUT017   
25                Household   45.9060            OUT017   
28                    Dairy   45.5086            OUT010   
...                     ...       ...               ...   
8502              Household  216.4192            OUT045   
8508  Fruits and Vegetables  199.4742            OUT045   
8509  Fruits and Vegetables  213.9902            OUT010   
8514                 Canned   57.5904            OUT045   
8519           Baking Goods  108.1570            OUT045   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
8                          2002         NaN               Tier 2   
9                          2007         NaN               Tier 2   
25                         2007         NaN               Tier 2   
28                         1998         NaN               Tier 3   
...                         ...         ...                  ...   
8502                       2002         NaN               Tier 2   
8508                       2002         NaN               Tier 2   
8509                       1998         NaN               Tier 3   
8514                       2002         NaN               Tier 2   
8519                       2002         NaN               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
3         Grocery Store           732.3800  
8     Supermarket Type1          1076.5986  
9     Supermarket Type1          4710.5350  
25    Supermarket Type1           838.9080  
28        Grocery Store           178.4344  
...                 ...                ...  
8502  Supermarket Type1          3020.0688  
8508  Supermarket Type1          2587.9646  
8509      Grocery Store           424.7804  
8514  Supermarket Type1           468.7232  
8519  Supermarket Type1           549.2850  

[2410 rows x 12 columns]

In [16]:
# Usamos describe para analizar la estadistica descriptiva de la columna
df['Outlet_Size'].describe()

count       6113
unique         3
top       Medium
freq        2793
Name: Outlet_Size, dtype: object

In [17]:
# Identificamos que valores existen en la columna
unique_values = df['Outlet_Size'].unique()
print("Valores únicos en 'Outlet_Size':")
print(unique_values)

Valores únicos en 'Outlet_Size':
['Medium' nan 'High' 'Small']


In [18]:
# agrupo para ver la relacion de ('Outlet_Type') y ('Outlet_Location_Type') en base al Outlet Size
df.groupby(['Outlet_Type', 'Outlet_Location_Type'])['Outlet_Size'].value_counts()

Outlet_Type        Outlet_Location_Type  Outlet_Size
Grocery Store      Tier 1                Small          528
Supermarket Type1  Tier 1                Medium         930
                                         Small          930
                   Tier 2                Small          930
                   Tier 3                High           932
Supermarket Type2  Tier 3                Medium         928
Supermarket Type3  Tier 3                Medium         935
Name: Outlet_Size, dtype: int64

In [19]:
# relación Outlet_Size mas comun para cada tipo de Outlet_Location_Type) .
df.groupby('Outlet_Location_Type')['Outlet_Size'].agg(pd.Series.mode)

Outlet_Location_Type
Tier 1     Small
Tier 2     Small
Tier 3    Medium
Name: Outlet_Size, dtype: object

In [20]:
# recuento Outlet_Size para cada Outlet_Identifier.
df.groupby(['Outlet_Identifier'])['Outlet_Size'].count()

Outlet_Identifier
OUT010      0
OUT013    932
OUT017      0
OUT018    928
OUT019    528
OUT027    935
OUT035    930
OUT045      0
OUT046    930
OUT049    930
Name: Outlet_Size, dtype: int64

In [21]:
# reemplazo valores faltantes (NaN) en el DataFrame df para la columna 'Outlet_Size'
df.loc[df['Outlet_Identifier'] == 'OUT010'] = df.loc[df['Outlet_Identifier'] == 'OUT010'].fillna('Medium')
df.loc[df['Outlet_Identifier'] == 'OUT017'] = df.loc[df['Outlet_Identifier'] == 'OUT017'].fillna('Small')
df.loc[df['Outlet_Identifier'] == 'OUT045'] = df.loc[df['Outlet_Identifier'] == 'OUT045'].fillna('Small')

In [22]:
df.loc[df['Outlet_Identifier'] == 'OUT017']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
9              FDU28       19.200          Regular         0.094450   
25             NCD06       13.000          Low Fat         0.099887   
53             FDA43       10.895          Low Fat         0.065042   
54             NCP18       12.150          Low Fat         0.028760   
73             FDG02        7.855          Low Fat         0.011325   
...              ...          ...              ...              ...   
8493           FDP21        7.420          Regular         0.025886   
8494           NCI54       15.200          Low Fat         0.000000   
8496           FDJ57        7.420          Regular         0.021696   
8500           NCQ42       20.350          Low Fat         0.000000   
8501           FDW21        5.340          Regular         0.005998   

                  Item_Type  Item_MRP Outlet_Identifier  \
9              Frozen Foods  187.8214            OUT017   
25                Household   45.9060            OUT017   
53    Fruits and Vegetables  196.3794            OUT017   
54                Household  151.4708            OUT017   
73                   Canned  189.6188            OUT017   
...                     ...       ...               ...   
8493            Snack Foods  189.1872            OUT017   
8494              Household  110.4912            OUT017   
8496                Seafood  185.3582            OUT017   
8500              Household  125.1678            OUT017   
8501            Snack Foods  100.4358            OUT017   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
9                          2007       Small               Tier 2   
25                         2007       Small               Tier 2   
53                         2007       Small               Tier 2   
54                         2007       Small               Tier 2   
73                         2007       Small               Tier 2   
...                         ...         ...                  ...   
8493                       2007       Small               Tier 2   
8494                       2007       Small               Tier 2   
8496                       2007       Small               Tier 2   
8500                       2007       Small               Tier 2   
8501                       2007       Small               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
9     Supermarket Type1          4710.5350  
25    Supermarket Type1           838.9080  
53    Supermarket Type1          3121.2704  
54    Supermarket Type1          4815.0656  
73    Supermarket Type1          2285.0256  
...                 ...                ...  
8493  Supermarket Type1          4727.1800  
8494  Supermarket Type1          1637.8680  
8496  Supermarket Type1          3715.1640  
8500  Supermarket Type1          1907.5170  
8501  Supermarket Type1          1508.0370  

[926 rows x 12 columns]

In [23]:
# valido si completé los valores faltantes
df['Outlet_Size'].value_counts()

Small     4243
Medium    3348
High       932
Name: Outlet_Size, dtype: int64

In [24]:
# check de datos faltantes
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

Ya no faltan datos.

Revisión consistencia tió de datos diccionario - data frame


Diccionario de datos

`Item_Identifier `- Identificador único del producto

`Item_Weight` - Peso del producto

`Item_Fat_Content` - Si el producto es bajo en grasa o regular

`Item_Visibility` - Porcentaje de área total de exposición de todos los productos para una tienda asignada al producto concreto

`Item_Type` - Categoría a la que el producto pertenece

`Item_MRP ` - Precio máximo de venta al público (precio de catálogo) del producto

`Outlet_Identifier `- Identificador único de la tienda

`Outlet_Establishment_Year` - Año en que se estableció la tienda

`Outlet_Size` - El tamaño de la tienda en relación al área total

`Outlet_Location_Type `- El tipo de área donde se encuentra la tienda

`Outlet_Type` - Si el punto de venta es una tienda de comestibles o algún tipo de supermercado

`Item_Outlet_Sales` - Ventas del producto en una tienda particular. Esta la variable objetivo a predecir.


In [26]:
# validamos info del data frame para checkear inconsistencias
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                8523 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                8523 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [28]:
df.nunique()

Item_Identifier              1559
Item_Weight                   416
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

In [29]:
#  Item_Fat_Content debe ser Low Fat o Regular, pero se presentan distintos valores, por lo que tratamos la data
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [30]:
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('reg', 'Regular')
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('LF', 'Low Fat')
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('low fat', 'Low Fat')
print(df['Item_Fat_Content'].value_counts())

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64


In [31]:
# estadisticas resumidas de columnas
df.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8523.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.226124         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       9.310000         0.026989    93.826500                1987.000000   
50%      12.857645         0.053931   143.012800                1999.000000   
75%      16.000000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800